pkg/document/crdt/element_rht.go

```diff
@@ -0,0 +1,211 @@
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package crdt
+
+import (
+	"fmt"
+	"sort"
+	"strings"
+
+	"github.com/yorkie-team/yorkie/pkg/document/time"
+)
+
+// ElementRHTNode is a node of ElementRHT.
+type ElementRHTNode struct {
+	key  string
+	elem Element
+}
+
+func newElementRHTNode(key string, elem Element) *ElementRHTNode {
+	return &ElementRHTNode{
+		key:  key,
+		elem: elem,
+	}
+}
+
+// Remove removes this node. It only marks the deleted time (tombstone).
+func (n *ElementRHTNode) Remove(removedAt *time.Ticket) bool {
+	if removedAt != nil && removedAt.After(n.elem.CreatedAt()) {
+		if n.elem.RemovedAt() == nil || removedAt.After(n.elem.RemovedAt()) {
+			return n.elem.Remove(removedAt)
+		}
+	}
+	return false
+}
+
+func (n *ElementRHTNode) isRemoved() bool {
+	return n.elem.RemovedAt() != nil
+}
+
+// Key returns the key of this node.
+func (n *ElementRHTNode) Key() string {
+	return n.key
+}
+
+// Element returns the element of this node.
+func (n *ElementRHTNode) Element() Element {
+	return n.elem
+}
+
+// ElementRHT is a hashtable with logical clock(Replicated hashtable).
+// The difference from RHT is that it keeps multiple values in one key. Using
+// Max Heap, the recently inserted value from the logical clock is returned
+// to the outside.
+type ElementRHT struct {
+	nodeMapByKey       map[string]*ElementRHTNode
+	nodeMapByCreatedAt map[string]*ElementRHTNode
+}
+
+// NewElementRHT creates a new instance of ElementRHT.
+func NewElementRHT() *ElementRHT {
+	return &ElementRHT{
+		nodeMapByKey:       make(map[string]*ElementRHTNode),
+		nodeMapByCreatedAt: make(map[string]*ElementRHTNode),
+	}
+}
+
+// Get returns the value of the given key.
+func (rht *ElementRHT) Get(key string) Element {
+	if node, ok := rht.nodeMapByKey[key]; ok {
+		if node.isRemoved() {
+			return nil
+		}
+		return node.elem
+	}
+	return nil
+}
+
+// Has returns whether the element exists of the given key or not.
+func (rht *ElementRHT) Has(key string) bool {
+	if node, ok := rht.nodeMapByKey[key]; ok {
+		return node != nil && !node.isRemoved()
+	}
+	return false
+}
+
+// Set sets the value of the given key. If there is an existing value, it is removed.
+func (rht *ElementRHT) Set(k string, v Element) Element {
+	node, ok := rht.nodeMapByKey[k]
+	var removed Element
+	if ok && !node.isRemoved() && v.CreatedAt().After(node.elem.CreatedAt()) {
+		node.Remove(v.CreatedAt())
+		removed = node.elem
+	}
```

Comment:
It seems that some of the logic is redundant. Is it possible to change it like the below?

```suggestion
	if ok && node.Remove(v.CreatedAt()) {
		removed = node.elem
	}
```